src/services/game-logic/GameEngine.ts

```diff
@@ -12,13 +13,59 @@ class GameEngine {
 
     getLegalMoves = (square: Square): Square[] => {
         const piece = this.board.getPiece(square);
-        let legalMoves = piece?.getPossibleMoves(this.board).filter(this.isOnBoard) ?? [];
+        let legalMoves = piece?.getPossibleMoves(this.board) ?? [];
         if (piece?.name === PieceNames.KING) {
             legalMoves = legalMoves.filter(
                 (move) => !this.isCastling(move, piece) || this.isCastlingLegal(move, piece)
             );
         }
-        return legalMoves.filter((destination) => !this.isOccupiedBySameColor(destination, piece));
+        return legalMoves
+            .filter((destination) => !this.isOccupiedBySameColor(destination, piece))
+            .filter((move) => this.moveNotResultWithCheck(move, piece, square));
+    };
+
+    private moveNotResultWithCheck(move: Square, piece: Piece | null, square: Square) {
+        const potentialMove: Square = { row: move.row, column: move.column };
+        const potentialPiece = this.board.getPiece(potentialMove);
+
+        this.movePiece(square, potentialMove);
+        piece!.hasMoved = true;
+        const isMovePossible = !this.isCheck(piece);
+
+        this.movePiece(potentialMove, square);
+        piece!.hasMoved = false;
```

Comment:
to tylko działa, jeśli bierka się wcześniej nie ruszyła - trzeba by zapisywać jej poprzedni status do stałej

Reply:
hmm, ale ona zawsze będzie się ruszać, o tutaj:
`this.movePiece(square, potentialMove);`